docs: Improve developer guide (#23669)
argoproj/argo-cd: docs/developer-guide/development-cycle.md

# The Development Cycle

## Prerequisites
1. [Development Environment](development-environment.md)   
2. [Toolchain Guide](toolchain-guide.md)

## Preface 

> When you have developed and possibly manually tested the code you want to contribute, you should ensure that everything will build correctly. Commit your changes to the local copy of your Git branch and perform the following steps, for each step the commands for both local and virtualized toolchain are available. 

Reviewer: ```suggestion
When you have developed and possibly manually tested the code you want to contribute, you should ensure that everything builds correctly. Commit your changes locally and perform the following steps, for each step the commands for both local and virtualized toolchain are listed. 
```

> ### Docker priviliges for virtualized chain users

Reviewer: ```suggestion
### Docker priviliges for virtualized toolchain users
```

[These instructions](toolchain-guide.md#docker-privileges) are relevant for most of the steps below 

> ### Using Podman for virtualized chain users

Reviewer: ```suggestion
### Using Podman for virtualized toolchain users
```

[These instructions](toolchain-guide.md#using-podman) are relevant for most of the steps below 

## Development cycle steps
### Set kubectl context to argocd namespace

> Set kubectl config to argocd namespace is required for all the following steps to succeed.

Reviewer: ```suggestion
Setting kubectl config to the argocd namespace is required for these steps to succeed.
```

All following commands in this guide assume the namespace is already set.

```shell
kubectl config set-context --current --namespace=argocd
```

### Pull in all build dependencies

As build dependencies change over time, you have to synchronize your development environment with the current specification. In order to pull in all required dependencies, issue:

* `make dep-ui` or `make dep-ui-local`

> Argo CD recently migrated to Go modules. Usually, dependencies will be downloaded on build time, but the Makefile provides two targets to download and vendor all dependencies:

Reviewer: ```suggestion
Argo CD recently migrated to Go modules. Usually, dependencies will be downloaded at build time, but the Makefile provides two targets to download and vendor all dependencies:
```

* `make mod-download` or `make mod-download-local` will download all required Go modules and
* `make mod-vendor` or `make mod-vendor-local` will vendor those dependencies into the Argo CD source tree

### Generate API glue code and other assets

Argo CD relies on Google's [Protocol Buffers](https://developers.google.com/protocol-buffers) for its API, and this makes heavy use of auto-generated glue code and stubs. Whenever you touched parts of the API code, you must re-generate the auto generated code.

* Run `make codegen` or `make codegen-local`, this might take a while
* Check if something has changed by running `git status` or `git diff`
* Commit any possible changes to your local Git branch, an appropriate commit message would be `Changes from codegen`, for example.

!!!note

>     There are a few non-obvious assets that are auto-generated. You should not change the autogenerated assets, as they will be overwritten by a subsequent run of `make codegen`. Instead, change their source files. Prominent examples of non-obvious auto-generated code are `swagger.json` or the installation manifest YAMLs.

Reviewer 2: ```suggestion
   There are a few non-obvious assets that are auto-generated. You should not edit these generated assets directly, as they will be overwritten by a subsequent run of make codegen. Instead, modify the source files from which they are generated such as Go types, annotations, or .proto files and then re-run the appropriate code generation command. Prominent examples of non-obvious auto-generated code are `swagger.json` or the installation manifest YAMLs.
```

> Instead, change their source files.

I was talking to @crenshaw-dev the other day about it, and he mentioned that it actually depends on whether you should be changing the source files or the proto files. The above note is true wrt to `swagger.json` that should never be edited manually. However, If a `.proto` file generates the auto-generated asset, changing the proto file makes sense here. If a Go struct with annotations (like Kubernetes CRD definitions) is the source, changing the Go source file makes sense. Hence, it actually depends.

Reviewer 2: Secondly, we should mention this detail about never changing the proto field number in case of field deprecation, as that may lead to API API-breaking change.

Author: I think we should deal with the codegen -related docs in a separate PR, this PR is more about restructuring stuff related to setting up the environment.

### Build your code and run unit tests

After the code glue has been generated, your code should build and the unit tests should run without any errors. Execute the following statements:

* `make build` or `make build-local`
* `make test` or `make test-local`

These steps are non-modifying, so there's no need to check for changes afterward.

### Lint your code base

In order to keep a consistent code style in our source tree, your code must be well-formed in accordance to some widely accepted rules, which are applied by a Linter.

The Linter might make some automatic changes to your code, such as indentation fixes. Some other errors reported by the Linter have to be fixed manually.

* Run `make lint` or `make lint-local` and observe any errors reported by the Linter
* Fix any of the errors reported and commit to your local branch

> * Finally, after the Linter reports no errors anymore, run `git status` or `git diff` to check for any changes made automatically by Lint

Reviewer: ```suggestion
* Finally, after the Linter reports no errors, run `git status` or `git diff` to check for any changes made automatically by Lint
```

* If there were automatic changes, commit them to your local branch

If you touched UI code, you should also run the Yarn linter on it:

* Run `make lint-ui` or `make lint-ui-local`
* Fix any of the errors reported by it

### Run end-to-end tests

> The final step is running the End-to-End testsuite, which makes sure that your Kubernetes dependencies are working properly. This will involve starting all the Argo CD components on your computer. The end-to-end tests consists of two parts: a server component, and a client component.

Reviewer: ```suggestion
The final step is running the End-to-End testsuite, which ensures that your Kubernetes dependencies are working properly. This will involve starting all the Argo CD components on your computer. The end-to-end tests consists of two parts: a server component, and a client component.
```
